File: Sources/UI/DataSourceDelegateObjC/CollectionView/CollectionDataSourceDelegate/CollectionDataSourceDelegate.m
//
//  CollectionDataSourceDelegate.m
//  Catalog
//
//  Created by Ivana Mršić on 20/05/2019.
//  Copyright © 2019 Infinum. All rights reserved.
//

#import "CollectionDataSourceDelegate.h"
#import "BaseCollectionSection.h"

@interface CollectionDataSourceDelegate ()

#pragma mark - Private properties

@property (nonatomic) UICollectionView *collectionView;
@property (nonatomic) id<CollectionViewReloader> reloader;

@end

@implementation CollectionDataSourceDelegate

#pragma mark - Inits

- (instancetype)initWithCollectionView:(UICollectionView *)collectionView
{
    return [self initWithCollectionView:collectionView reloader:[DefaultCollectionViewReloader new]];
}

- (instancetype)initWithCollectionView:(UICollectionView *)collectionView reloader:(nonnull id<CollectionViewReloader>)reloader
{
    self = [super init];
    if (self) {
        _collectionView = collectionView;
        _reloader = reloader;
        
        _collectionView.delegate = self;
        _collectionView.dataSource = self;
    }
    return self;
}

#pragma mark - Setters

- (void)setSections:(NSArray<id<CollectionSectionItem>> *)sections
{
    NSArray *oldSections = _sections;
    
    _sections = sections;
    
    [self.reloader reloadCollectionView:self.collectionView oldSections:oldSections newSections:sections];
}

- (void)setItems:(NSArray<id<CollectionCellItem>> *)items
{
    self.sections = @[[[BaseCollectionSection alloc] initWithItems:items]];
}

#pragma mark - Getters

- (NSArray<id<CollectionCellItem>> *)items
{
    NSMutableArray<id<CollectionCellItem>> *items = [NSMutableArray new];
    
    for (id<CollectionSectionItem> section in self.sections) {
        if (!section.items) { break; }
        
        [items addObjectsFromArray:section.items];
    }
    
    return items;
}

#pragma mark - UICollectionViewDataSource methods

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
{
    return self.sections.count;
}

- (NSInteger)collectionView:(nonnull UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return [self.sections[section] items].count;
}

- (nonnull __kindof UICollectionViewCell *)collectionView:(nonnull UICollectionView *)collectionView cellForItemAtIndexPath:(nonnull NSIndexPath *)indexPath {
    return [self.sections[indexPath.section].items[indexPath.row] cellFromCollectionView:collectionView at:indexPath];
}

#pragma mark - UICollectionViewDelegate methods

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    [self.collectionView deselectItemAtIndexPath:indexPath animated:YES];
    [self.sections[indexPath.section].items[indexPath.row] didSelectAt:indexPath];
}

- (void)collectionView:(UICollectionView *)collectionView willDisplayCell:(UICollectionViewCell *)cell forItemAtIndexPath:(NSIndexPath *)indexPath
{
    [self.sections[indexPath.section].items[indexPath.row] willDisplayCell:cell at:indexPath];
}

#pragma mark - UICollectionViewDelegateFlowLayout

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath
{
    NSInteger sectionsCount = self.sections[indexPath.section].items.count;
    
    return [self.sections[indexPath.section].items[indexPath.row] itemSizeForCollectionViewSize:collectionView.bounds.size layout:collectionViewLayout at:indexPath numberOfItemsInSection:sectionsCount];
}

@end

File: Sources/UI/DataSourceDelegate/Swift/Collection View/CollectionDataSourceDelegate.swift
//
//  CollectionDataSourceDelegate.swift
//
//  Created by Vlaho Poluta
//  Copyright © 2019 Infinum. All rights reserved.
//

import UIKit
import RxSwift
import RxCocoa

/// Object serving as a data source and delegate for a collection view.
/// Main purpose is to reduce a boilerplate when dealing with simple
/// collection view data displaying.
///
/// It can handle both - sections and items
public class CollectionDataSourceDelegate: NSObject {
    
    // MARK: - Public properties

    /// Setting a sections will invoke internal reloader causing table view to refresh.
    public var sections: [CollectionSectionItem]? {
        didSet(oldValue) {
            reloader.reload(collectionView, oldSections: oldValue, newSections: sections)
        }
    }
    
    /// Setting an items will invoke internal reloader causing table view to refresh.
    ///
    /// If there are multiple sections - then data is flattened to single array
    public var items: [CollectionCellItem]? {
        get {
            return sections?
                .map(\.items)
                .reduce(into: [CollectionCellItem]()) { $0 = $0 + $1 }
        }
        set {
            let section: CollectionSectionItem? = BlankCollectionSection(items: newValue)
            sections = [section].compactMap { $0 }
        }
    }
    
    // MARK: - Private properties

    private let collectionView: UICollectionView
    private let reloader: CollectionViewReloader
    private let disposeBag = DisposeBag()
    
    /// Creates a new data source delegate object responsible for handling
    /// collection view's data source and delegate logic.
    ///
    /// **If using the data source delegate object do not change the table view
    /// dataSource property since this object depends on it**
    ///
    /// Freely use `delegate` property since internally data source delegate will
    /// use pass through delegate.
    ///
    /// - Parameters:
    ///   - collectionView: Collection view to control
    ///   - reloader: Data reloader
    public init(collectionView: UICollectionView, reloader: CollectionViewReloader = DefaultCollectionViewReloader()) {
        self.collectionView = collectionView
        self.reloader = reloader
        super.init()
        
        collectionView.dataSource = self
        collectionView.rx
            .setDelegate(self)
            .disposed(by: disposeBag)
    }
    
}

extension CollectionDataSourceDelegate: UICollectionViewDataSource {
    
    public func numberOfSections(in collectionView: UICollectionView) -> Int {
        return sections?.count ?? 0
    }
    
    public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return sections?[section].items.count ?? 0
    }
    
    public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        return sections![indexPath].cell(from: collectionView, at: indexPath)
    }
    
}

extension CollectionDataSourceDelegate: UICollectionViewDelegate {
    
    public func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        collectionView.deselectItem(at: indexPath, animated: true)
        sections?[indexPath].didSelect(at: indexPath)
    }
    
    public func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
        sections?[indexPath].willDisplay(cell: cell, at: indexPath)
    }
    
}

extension CollectionDataSourceDelegate: UICollectionViewDelegateFlowLayout {
    
    public func collectionView(
        _ collectionView: UICollectionView,
        layout collectionViewLayout: UICollectionViewLayout,
        sizeForItemAt indexPath: IndexPath
    ) -> CGSize {
        let sectionsCount = sections![indexPath.section].items.count
        
        return sections![indexPath]
            .itemSize(
                for: collectionView.bounds.size,
                layout: collectionViewLayout as? UICollectionViewFlowLayout,
                at: indexPath,
                numberOfItemsInSection: sectionsCount
            )
    }
    
}
